### ISSUE TITLE
Use of scipy to replace core logic in solveGfe routine

### ISSUE THREAD

**Issue by quinnreynolds** [quinnreynolds/minplascalc#16]
**[Original report](https://bitbucket.org/kittychunk/minplascalc/issue/15) by Quinn Reynolds (Bitbucket: [kittychunk](https://bitbucket.org/kittychunk), GitHub: [kittychunk](https://github.com/kittychunk)).**

----------------------------------------

Open to suggestions to replace the custom minimiser routine used in solveGfe with something more robust. Some attempts were made with scipy.optimize during initial development, but the aggressive nonlinearity of the Gibbs free energy function and the wide ranges in magnitude of the ni variables make it challenging for the standard algorithms. However, someone with more scipy experience than me is welcome to take a stab at it.

**Comment by alchemyst** [comment]
I profiled and instrumented the minimiser to see whether scipy would actually help here. Summary: the custom Newton step is not the weak part, and the thing that *is* weak is a one-line fix rather than a library swap.

## The Newton step is sound

`gfe_matrix[:nb, :nb]` is built as `-kT/N_tot + diag(kT/N_i)`, and since
`μ_i = E⁰_i - kT[ln N_tot + ln(z_i kT/P) - ln N_i]`,

```
∂μ_i/∂N_j = -kT/N_tot + δ_ij · kT/N_i
```

which is exactly that block. The bordered system with the stoichiometry and charge constraints is the KKT system for the equality-constrained Newton step. So the exact gradient and Hessian are already there in closed form, and the linear solve is **0.8% of composition runtime**. There is nothing for a general-purpose optimiser to improve on in the step itself.

(One caveat: `E0`/`dE` depend on the composition and are refreshed each iteration but not differentiated through, so it is strictly a quasi-Newton on that coupling.)

## The step control is the weak part

The governor caps the change in **every** species at `governor_factor * N_i`, **symmetrically**:

```python
max_allowed_delta_Ni = governor_factor * self.__Ni
delta_Ni = delta_Ni.clip(min=max_allowed_delta_Ni)
relaxation_factor = (max_allowed_delta_Ni / delta_Ni).min()
```

With `governor_factor <= 0.9` a species can grow by at most a factor of 1.9 per iteration, and the `.min()` runs over all species, so one species needing a large relative move throttles the entire step. A trace ion climbing several decades through an ionisation front is rate-limited by `log(range)/log(1.9)`.

Only the *decrease* direction actually threatens the solve — `N_i` must stay positive for the `log` and the `1/N_i` Hessian entries. Replacing the symmetric cap with the fraction-to-boundary rule from interior-point methods (limit only steps that would drive a species non-positive; leave growth free), keeping the same Newton step and the same convergence test:

| step control | converged | iterations, converged states |
|---|---|---|
| governor (current) | 143 / 168 | baseline |
| fraction-to-boundary | **156 / 168** | **1.5–1.9x fewer** |

over 168 (composition, pressure, temperature) states spanning 500–30000 K and 1 kPa–1 MPa. That recovers about half the outright failures. The remaining failures are all high-temperature, low-pressure.

## The convergence test is a separate, real defect

`mixture.py` stops on the relative movement of the **most abundant species only** (there is already a `TODO` there). At 2000 K that leaves the electron density **24% from converged**, which propagates to a **12.5% error in electrical conductivity**. Above ~4000 K it is negligible.

Worth flagging because the obvious fix does not work: widening the test to all species converges only **80 / 168**, regardless of step control. Trace species sitting at 1e-183 mole fraction cannot be resolved to 1e-10 *relative* in double precision. This needs a scaled or absolute-floored criterion, not a naive widening.

## On scipy specifically

It is perfectly possible to hand the analytic Hessian over — `trust-constr` takes `jac`, `hess`, `LinearConstraint` and bounds; `root` takes an analytic Jacobian. But the measurements say the globalisation is what needs fixing, and that is achievable in a few lines without a dependency or a rewrite.

If scipy is revisited later, the thing that made the original attempts hard is worth naming: it is the **dynamic range of `n_i`** (mole fractions spanning ~180 decades at low temperature), not the nonlinearity of G. A generic trust region faces exactly the same problem. The reformulation that actually addresses it is solving for `y_i = ln n_i` — positivity becomes automatic and a 180-decade multiplicative range becomes a modest additive one. That is a bigger change, and it would be the right prerequisite for any scipy-based attempt.

## Suggested order

1. Fraction-to-boundary step control — one line, 156/168, 1.5–1.9x fewer iterations.
2. A properly scaled convergence criterion — fixes a real 12.5% error in σ at low T. Needs design, not a one-liner.
3. Warm-start across temperature. A sweep currently restarts from `gfe_initial_particles` at every T, discarding an excellent initial guess; the low-T end costs 182 Newton iterations against 24 at high T.
4. Only then consider scipy, and probably only together with the log-space reformulation.

Prototypes and the measurement scripts are on the `feature/optimize-notation-and-computation` branch (`bench/minimiser_variants.py`, `bench/check_minimiser.py`, `bench/check_minimiser_variants.py`). No `src/` changes for any of this yet.

**Comment by quinnreynolds** [comment]
Thanks Carl and Claude :-), nice analysis and it seems to suggest we shouldn't spend too much time on (or have too many expectations from) moving the minimiser to a scipy standard.

Just one comment on point 3: Warm start across temperature was actually part of the original behaviour in the early days, but we abandoned it because the change limiters restrict how much a value can change in any given iteration - in cases where a species has converged down to a low value (1e-100 or lower) in the previous iteration but is now present in significant quantities, it could take a *very* long time for it to climb back up. This can happen when e.g. going from a low temperature to a high one, or crossing some critical boundary in the physics such as the onset of ionisation. We should just make sure that the new warm start implementation doesn't unintentionally impair convergence in cases like that.

One way to handle this might be to have explicit functions for generating temperature/pressure sweeps, which could then use some custom start-condition logic internally?

**Comment by quinnreynolds** [comment]
One additional general comment here: In GFE minimisers for other complex systems (metallurgy is a good example), it does seem like it's quite common to cast the problem into log space and solve it there. That's a good suggestion to explore anyway, even if we don't go the scipy route.

**Comment by alchemyst** [comment]
In this most recent round of work I found myself wanting a more robust benchmarking suite. @quinnreynolds I wonder if you could contribute some of the real-world applications of minplascalc that will keep benchmarking honest? As you say at the moment the docs' sweep is so dominant in time that it might bias us in that direction.

Also agree on solving in log space. I like both the positivity guarantees it makes as well as the scaling benefits.

**Comment by quinnreynolds** [comment]
Agreed Carl. Two possible ways to approach extending the benchmark/test suite:

1. I could add a few extra species to the small database that's part of the minplascalc repo, and set up additional mixture calculation unit tests. Pro: single, open test target packaged with the repo. Con: need to be cognizant of the limits that NIST place on repackaging their data, test suite could grow large and unwieldy if we continually add new cases, issue-specific unit tests become redundant after their associated issues are resolved and closed.
2. I could share some of our real-world mixture calculations with you directly as a private test suite. Pro: can share issue-specific test cases without polluting repo, can investigate more complex test and edge cases. Con: separate target set not visible to general users.

Let me know how you'd prefer to go ahead here?

**Comment by alchemyst** [comment]
You're right to point to the IP component of this. I think we should keep things as unencumbered as possible. Is there a third way where we just add some more examples that use the same components we have now but exercise some other part of the code? 

Either way this feels like it should be an issue by itself. I'll create it now.

**Comment by alchemyst** [comment]
Benchmarking/performance suite moved to #94
